modules/fedimint-wallet-client/src/backup.rs

```diff
@@ -40,3 +60,377 @@ pub struct WalletModuleBackupV0 {
+        };
+        Ok(dbtx.get_value(&RecoveryStateKey)
+            .await
+            .and_then(|(state, common)| {
+                if let WalletRecoveryState::V0(state) = state {
+                    Some((state, common))
+                } else {
+                    warn!(target: LOG_CLIENT_RECOVERY, "Found unknown version recovery state. Ignoring");
+                    None
+                }
+            })
+            .map(|(state, common)| {
+                (
+                    WalletRecovery {
+                        state,
+                        data,
+                        btc_rpc,
+                    },
+                    common,
+                )
+            }))
+    }
+
+    async fn store_dbtx(
+        &self,
+        dbtx: &mut DatabaseTransaction<'_>,
+        common: &RecoveryFromHistoryCommon,
+    ) {
+        trace!(target: LOG_CLIENT_MODULE_WALLET, "Storing recovery state");
+        dbtx.insert_entry(
+            &RecoveryStateKey,
+            &(WalletRecoveryState::V0(self.state.clone()), common.clone()),
+        )
+        .await;
+    }
+
+    async fn delete_dbtx(&self, dbtx: &mut DatabaseTransaction<'_>) {
+        dbtx.remove_entry(&RecoveryStateKey).await;
+    }
+
+    async fn load_finalized(dbtx: &mut DatabaseTransaction<'_>) -> Option<bool> {
+        dbtx.get_value(&RecoveryFinalizedKey).await
+    }
+
+    async fn store_finalized(dbtx: &mut DatabaseTransaction<'_>, state: bool) {
+        dbtx.insert_entry(&RecoveryFinalizedKey, &state).await;
+    }
+
+    async fn handle_input(
+        &mut self,
+        _client_ctx: &ClientContext<WalletClientModule>,
+        _idx: usize,
+        input: &WalletInput,
+        session_idx: u64,
+    ) -> anyhow::Result<()> {
+        let script_pubkey = match input {
+            WalletInput::V0(WalletInputV0(ref input)) => &input.tx_output().script_pubkey,
+            WalletInput::Default {
+                variant: _,
+                bytes: _,
+            } => {
+                return Ok(());
+            }
+        };
+
+        self.state
+            .tracker
+            .handle_script(&self.data, script_pubkey, session_idx);
+
+        Ok(())
+    }
+
+    async fn pre_finalize(&mut self) -> anyhow::Result<()> {
+        let data = &self.data;
+        let btc_rpc = &self.btc_rpc;
+        // Due to lifetime in async context issue, this one is cloned and wrapped in a
+        // mutex
+        let tracker = &Arc::new(Mutex::new(self.state.tracker.clone()));
+
+        debug!(target: LOG_CLIENT_MODULE_WALLET,
+            next_unused_tweak_idx = ?self.state.next_unused_idx_from_backup,
+            "Scanning blockchain for used peg-in addresses");
+        let RecoverScanOutcome { last_used_idx: _, new_start_idx, tweak_idxes_with_pegins}
+            = recover_scan_idxes_for_activity(
+                self.state.next_unused_idx_from_backup,
+                self.state.tracker.used_tweak_idxes(),
```

Comment:
Eventually we should scan from index zero to not miss anything, but we need the already claimed list in the backup to do so privately.